File: dynamic_listings/lua/default/containers/timer.lua
-- This is a simple timer class that executes a function after
-- a certain amount of time
Timer = {}

function Timer:new(o)
    o = o or {
        time = 0,
        set_time = 0,
        function_to_execute = nil,
        one_shot = false,
        active = false
    }
    setmetatable(o, self)
    self.__index = self
    return o
end

function Timer:update(dt)
    if (not self.active) then
        -- We return directly if the timer is disabled
        return
    end
    -- Like any other entity, we update it
    self.time = self.time - dt
    -- When the timer "ticks", we execute the function
    if (self.time <= 0) then
        self.function_to_execute()
        if (self.one_shot) then
            -- If this timer is a one-shot, we disable it
            self.active = false
        end
        -- We reset the timer (we may need to re-activate it manually later)
        self.time = self.set_time
    end
end

File: dynamic_listings/C++/default/containers/timer.cpp
#include <functional>

class Timer{
    /*
     * This is a simple timer class that executes a function after
     * a certain amount of time
     */
    private:
        float time;
        float set_time;
        std::function<void()> function_to_execute;
        bool one_shot;
        bool active;

    public:
        Timer(float time_set, function_ptr funct, bool oneshot = false, bool act=false){
            // We prepare the timer and memorize the setting
            time = time_set;
            set_time = time;
            // The function pointer should already be prepared with the arguments
            function_to_execute = funct;
            // Is this timer one-shot then disable?
            one_shot = oneshot;
            // Does this timer need to be active when constructed?
            active = act;
        }

        void update(float dt){
            if (!active){
                // We return directly if the timer is disabled
                return;
            }
            // Like any other entity, we update it
            time -= dt;
            // When the timer "ticks", we execute the function
            if (time <= 0){
                function_to_execute();
                if (one_shot){
                    // If this timer is a one-shot, we disable it
                    active = false;
                }
                // We reset the timer (we may need to re-activate it manually later)
                time = set_time;
            }
        }
};
